### PR TITLE
GUA-748- Query Activty Log

### PR DESCRIPTION
**Pull request by huwd**
Adds a Lambda that:

Reads in an from the dummy dynamoDb stream which is a TxMA event, and  queries the activity_log table to get an entry for the user_id and a session_id . 
The response from the dynamo db could be empty in which case the lambda sends just the TXMA event.
If a response id received both the response and the TXMA event is returned back to the next lambda.
The output is written  to a subsequent SQS queue

Note: There is a concern regarding the security implications of the data stored on the raw events store. Hence we are using a dummy event store for now. Once the concerns are addressed we can update the query lambda to read from the raw events store stream.

### CODE REVIEW

File: lambda/query-activity-log/models.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
+type UrnFdnSub = string;
```
Comment (DougsCER) [inline lambda/query-activity-log/models.ts:1]: why are some fields pulled out and others not?

File: lambda/query-activity-log/tests/query-activity-log.test.ts [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+            },
+          },
+        ],
+      })
+    );
+    expect(() => {
+      validateTxmaEventBody(txmaEvent);
+    }).toThrowError();
+  });
+  test("throws error when timestamp is missing", () => {
```
Comment (DougsCER) [inline lambda/query-activity-log/tests/query-activity-log.test.ts:106]: all the test files i've seen we have space above each test and each describe.  I think it helps readability a lot

File: lambda/query-activity-log/tests/query-activity-log.test.ts [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+        ],
+      })
+    );
+    expect(() => {
+      validateTxmaEventBody(txmaEvent);
+    }).toThrowError();
+  });
+  test("throws error when user_id  is missing", () => {
+    const txmaEvent = JSON.parse(
+      JSON.stringify({
```
Comment (DougsCER) [inline lambda/query-activity-log/tests/query-activity-log.test.ts:160]: This is ok but to be quicker, you can use object deconstruction and overrides to create tests parameters with altered fields 
Eg https://github.com/alphagov/di-account-management-backend/blob/main/lambda/save-raw-events/tests/save-raw-events.test.ts#L77


the syntax is 

`...<original object>, <field to mutate> : <new value>`


This means if the original object changes you don't need to re-write all the tests to update the object being created

File: lambda/query-activity-log/tests/query-activity-log.test.ts [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+    }).toThrowError();
+  });
+});
+
+describe("sendSqsMessage", () => {
+  beforeEach(() => {
+    sqsMock.reset();
+    process.env.QUEUE_URL = queueUrl;
+  });
+  afterEach(() => {
```
Comment (DougsCER) [inline lambda/query-activity-log/tests/query-activity-log.test.ts:209]: spacing again, personally i like spacing between functions for readability

File: lambda/shared-models.ts
Comment (DougsCER) [inline lambda/shared-models.ts:1]: this file will be removed by other PR. 👍

File: lambda/query-activity-log/models.ts (hunk 1 above)
Comment (divyaparameswaran) [reply lambda/query-activity-log/models.ts:1]: That's a good point not sure why it is needed . Happy for it to be all in the same format .

File: lambda/query-activity-log/tests/query-activity-log.test.ts (hunk 2 above)
Comment (divyaparameswaran) [reply lambda/query-activity-log/tests/query-activity-log.test.ts:106]: I assumed my prettier plugin would take care of it . I will add the space .

File: lambda/query-activity-log/models.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+  event_type: string;
+  session_id: string;
+  user_id: string;
+  timestamp: number;
+  activities: Activity[];
+  truncated: boolean;
+}
+
+export interface UserActivityLog {
+  TxmaEvent: TxmaEvent;
```
Comment (DougsCER) [inline lambda/query-activity-log/models.ts:36]: type parameters should be lower case i think

Comment (divyaparameswaran) [reply lambda/query-activity-log/models.ts:36]: yup okay I will fix this here and we can add ticket for all the other lambdas as well .

File: lambda/query-activity-log/query-activity-log.ts [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+  const { AWS_REGION } = process.env;
+  const client = new SQSClient({ region: AWS_REGION });
+  const message: SendMessageRequest = {
+    QueueUrl: queueUrl,
+    MessageBody: messageBody,
+  };
+  const result = await client.send(new SendMessageCommand(message));
+  return result.MessageId;
+};
+export const handler = async (event: DynamoDBStreamEvent): Promise<void> => {
```
Comment (DougsCER) [inline lambda/query-activity-log/query-activity-log.ts:93]: space

**Review by DougsCER — APPROVED** [review]

**Pull request merged**